### PR TITLE
New version of encrypt

### PR DESCRIPTION
**Pull request by justin1121**
Allows customizing the RSA and data encryption key passed into the encrypt function. Passing the data encryption key allows the user more control over what happens with the dek. The other component here is that the dek is cached by default. If encrypting multiple items it will reuse the same dek for efficiency reasons. Encrypting a bunch of things with RSA can be slow.

### CODE REVIEW

File: packages/cape/src/methods.ts [hunk 1]
```diff
@@ -50,6 +53,11 @@ interface Message {
   error: string;
 }
 
+interface EncryptOptions {
+  key?: string;
+  dek?: string;
```
Comment (gavinuhma) [inline packages/cape/src/methods.ts:58]: I like `dataKey` more than `dek` as a name

File: packages/cape/src/methods.ts [hunk 2]
```diff
@@ -61,6 +69,7 @@ export abstract class Methods {
   websocket?: WebsocketConnection;
   nonce?: string;
   checkDate?: Date;
+  savedDataEncryptedKey?: DataEncryptionKey;
```
Comment (gavinuhma) [inline packages/cape/src/methods.ts:72]: `savedDataKey`

File: packages/cape/src/methods.ts [hunk 3]
```diff
@@ -223,26 +232,51 @@ export abstract class Methods {
...
+      throw Error('cannot pass key and dek to this function');
+    }
+
+    if (options.dek != null) {
+      // TODO set this.saveDataEncryptedkey or something...
+      throw Error('not implemented');
     }
 
-    return await capeEncrypt(k, input);
+    if (this.savedDataEncryptedKey == null) {
```
Comment (gavinuhma) [inline packages/cape/src/methods.ts:273]: It seems we are just storing one data key, but we talked about having a key/value based on the rsa public key. In other words, we would create and store one data key for each public key

File: packages/cape/src/methods.ts [hunk 4]
```diff
@@ -223,26 +232,51 @@ export abstract class Methods {
...
+      this.savedDataEncryptedKey = await this.generateDataEncryptionKey(options.key);
+    }
```
Comment (gavinuhma) [inline packages/cape/src/methods.ts:275]: `generateDataKey`

File: packages/cape/src/encrypt.ts [hunk 5]
```diff
@@ -2,9 +2,14 @@ import { Aead, CipherSuite, Kdf, Kem } from 'hpke-js';
 import { debug } from 'loglevel';
 import * as forge from 'node-forge';
 
+export interface DataEncryptionKey {
```
Comment (gavinuhma) [inline packages/cape/src/encrypt.ts:5]: `DataKey`

**Review by gavinuhma — CHANGES_REQUESTED** [review]

File: packages/cape/src/methods.ts (hunk 3 above)
Comment (justin1121) [reply packages/cape/src/methods.ts:273]: Couldn't remember what we decided there works for me. Thanks!

File: packages/cape/src/methods.ts [hunk 6]
```diff
@@ -223,26 +231,61 @@ export abstract class Methods {
...
-  public async encrypt(input: string, key?: string): Promise<string> {
-    let k = key;
-    if (k == null) {
-      k = await this.key();
+  public async encrypt(
+    input: string,
+    options: EncryptOptions = { key: undefined, dataKey: undefined },
+  ): Promise<string> {
+    if (options.key != null && options.dataKey != null) {
+      throw Error('cannot pass key and dek to this function');
```
Comment (gavinuhma) [inline packages/cape/src/methods.ts:276]:
```suggestion
      throw Error('cannot pass key and datakey to this function');
```

File: packages/cape/src/encrypt.spec.ts [hunk 7]
```diff
@@ -11,24 +11,21 @@ describe('encrypt', () => {
   test('encrypt and decrypt the text', async () => {
     const text = 'hello world';
     const keyPair = await suite.generateKeyPair();
-    const { cipherText, encapsulatedKey } = await encrypt(encoder.encode(text), (keyPair.publicKey as XCryptoKey).key);
-    const plainText = await suite.open({ recipientKey: keyPair, enc: encapsulatedKey }, cipherText);
+    const { cipherText, plaintextDek } = await encrypt(encoder.encode(text), (keyPair.publicKey as XCryptoKey).key);
```
Comment (gavinuhma) [inline packages/cape/src/encrypt.spec.ts:14]:
```suggestion
    const { cipherText, plaintextDataKey } = await encrypt(encoder.encode(text), (keyPair.publicKey as XCryptoKey).key);
```
```suggestion
    const { cipherText, plaintextDataKey } = await encrypt(encoder.encode(text), (keyPair.publicKey as XCryptoKey).key);
```

File: packages/cape/src/encrypt.spec.ts [hunk 8]
```diff
@@ -11,24 +11,21 @@ describe('encrypt', () => {
...
+    const plainText = await suite.open({ recipientKey: keyPair, enc: plaintextDek }, cipherText);
```
Comment (gavinuhma) [inline packages/cape/src/encrypt.spec.ts:15]:
```suggestion
    const plainText = await suite.open({ recipientKey: keyPair, enc: plaintextDataKey }, cipherText);
```

File: packages/cape/src/encrypt.ts [hunk 9]
```diff
@@ -2,9 +2,14 @@ import { Aead, CipherSuite, Kdf, Kem } from 'hpke-js';
...
 
+export interface DataKey {
+  plaintext: string;
+  ciphertext: string;
+}
+
 interface EncryptResponse {
   cipherText: Uint8Array;
-  encapsulatedKey: Uint8Array;
+  plaintextDek: Uint8Array;
```
Comment (gavinuhma) [inline packages/cape/src/encrypt.ts:12]:
```suggestion
  plaintextDataKey: Uint8Array;
```

File: packages/cape/src/encrypt.ts [hunk 10]
```diff
@@ -28,7 +33,7 @@ export async function encrypt(plainText: Uint8Array, publicKey: Uint8Array): Pro
   // Encrypt the plain text using the public key
   const { ct, enc } = await suite.seal({ recipientPublicKey: rkp }, plainText);
   // Convert both the ct and enc from an ArrayBuffer to Uint8Array's.
-  return { cipherText: new Uint8Array(ct), encapsulatedKey: new Uint8Array(enc) };
+  return { cipherText: new Uint8Array(ct), plaintextDek: new Uint8Array(enc) };
```
Comment (gavinuhma) [inline packages/cape/src/encrypt.ts:36]:
```suggestion
  return { cipherText: new Uint8Array(ct), plaintextDataKey: new Uint8Array(enc) };
```

File: packages/cape/src/methods.ts [hunk 11]
```diff
@@ -165,8 +173,8 @@ export abstract class Methods {
       throw new Error('Unable to invoke the function, missing public key. Call Cape.connect() first.');
     }
     try {
-      const { cipherText, encapsulatedKey } = await encrypt(getBytes(data), this.publicKey);
-      const input = concat(encapsulatedKey, cipherText);
+      const { cipherText, plaintextDek } = await encrypt(getBytes(data), this.publicKey);
```
Comment (gavinuhma) [inline packages/cape/src/methods.ts:176]:
```suggestion
      const { cipherText, plaintextDataKey } = await encrypt(getBytes(data), this.publicKey);
```

File: packages/cape/src/methods.ts [hunk 12]
```diff
@@ -165,8 +173,8 @@ export abstract class Methods {
...
+      const input = concat(plaintextDek, cipherText);
```
Comment (gavinuhma) [inline packages/cape/src/methods.ts:177]:
```suggestion
      const input = concat(plaintextDataKey, cipherText);
```

File: packages/cape/src/methods.ts [hunk 13]
```diff
@@ -223,26 +231,61 @@ export abstract class Methods {
...
+    return newDataKey;
+  }
+
   /**
-   * Encrypt the
-   * The returned key is stored in the `client` object
+   * Encrypt the input.
+   *
+   * Encrypt takes options which allow you customize what RSA key or
+   * data encryption key (dek) is used. You cannot pass a custom RSA key
```
Comment (gavinuhma) [inline packages/cape/src/methods.ts:259]:
```suggestion
   * dataKey is used. You cannot pass a custom RSA key
```
```suggestion
   * dataKey is used. You cannot pass a custom RSA key
```

File: packages/cape/src/methods.ts [hunk 14]
```diff
@@ -223,26 +231,61 @@ export abstract class Methods {
...
+   * and dek. This function will return an error if you do.
```
Comment (gavinuhma) [inline packages/cape/src/methods.ts:260]:
```suggestion
   * and a . This function will return an error if you do.
```
```suggestion
   * and a dataKey. This function will return an error if you do.
```

**Review by gavinuhma — CHANGES_REQUESTED** [review]

**Pull request merged**